Implement User-Agent Client Hints - navigator.userAgentData

colelao opened the pull request:
#### 5bfcbdd6b8303a9df35caa888c268c9fe3710568
<pre>
Implement User-Agent Client Hints - navigator.userAgentData
<a href="https://bugs.webkit.org/show_bug.cgi?id=241749">https://bugs.webkit.org/show_bug.cgi?id=241749</a>
<a href="https://rdar.apple.com/95454627">rdar://95454627</a>

Reviewed by Brent Fulgham.

This patch populates the userAgentData data structure to be used for quirking. It follows this spec:
<a href="https://wicg.github.io/ua-client-hints/#getHighEntropyValues">https://wicg.github.io/ua-client-hints/#getHighEntropyValues</a>

Sensitive information is denied such as model and formFactors.

* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::userAgentData const):
(WebCore::Navigator::initializeNavigatorUAData const): Deleted.
* Source/WebCore/page/Navigator.h:
* Source/WebCore/page/NavigatorUAData.cpp:
(WebCore::NavigatorUAData::brands const):
(WebCore::NavigatorUAData::mobile const):
(WebCore::NavigatorUAData::platform const):
(WebCore::NavigatorUAData::toJSON const):
(WebCore::NavigatorUAData::getHighEntropyValues const):
(WebCore::NavigatorUAData::createArbitraryVersion):
(WebCore::NavigatorUAData::createArbitraryBrand):
* Source/WebCore/page/NavigatorUAData.h:
* Source/WebCore/page/UADataValues.h:
(WebCore::UADataValues::create):
(WebCore::UADataValues::UADataValues):
* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::userAgentData const):
(WebCore::WorkerNavigator::initializeNavigatorUAData const): Deleted.
* Source/WebCore/page/WorkerNavigator.h:

Canonical link: <a href="https://commits.webkit.org/298242@main">https://commits.webkit.org/298242@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d78c3e19ab7e6bb9fe485e1028fc56fb2d5dfc0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114630 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34376 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24839 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120796 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65352 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116519 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35004 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42936 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87135 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/42039 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117578 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27886 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102943 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67523 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27073 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21067 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64473 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97266 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21177 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123998 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41642 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31096 "Found 1 new test failure: fast/block/transparent-outline-with-and-without-border-radius.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95949 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42019 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99128 "") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95730 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40892 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18734 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37741 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18377 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41520 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47032 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41106 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44413 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42856 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -40,31 +56,116 @@ Ref<NavigatorUAData> NavigatorUAData::create()
+#else
+            values->architecture = "arm"_s;
+#endif
+        else if (hint == "bitness")
+            values->bitness = "64"_s;
```
Instead of lying, we _could_ provide the truth using `CPU(ADDRESS64)`.

Alternatively, we could add a comment to explain why it's worth lying about CPU bitness. I'm not sure it's really needed, though?

colelao replied:
cc @brentfulgham 

Is there any reason to lie about bitness?

brentfulgham replied:
Some watches are still 32-bit, so allowing a website to know this information is at least 1 bit of entropy. I don't see a compelling need for this to be shared, so I suggest we just stick to our most common configuration.

brentfulgham replied:
In fact, for now we may not want to supply any of the high-entropy information unless we identify a compelling reason to share the state.

We are only implementing these bits now to support Quirking sites that are ignoring the standards bodies and are requiring it for key site functions.

brentfulgham replied:
Although whatever already appears in the UA string should be provided through this interface, too, and using the same contents.

mcatanzaro commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -40,31 +56,116 @@ Ref<NavigatorUAData> NavigatorUAData::create()
+                struct utsname osname;
+                version.construct(uname(&osname) >= 0 ? makeString(unsafeSpan(osname.version)) : emptyString());
+            });
+            values->platformVersion = version->isolatedCopy();
```
This is not OK. Giving out the Linux kernel version is (a) not useful, and (b) major fingerprinting vector. This is too intense even for the high-entropy values.

I would just always hardcode a fake answer like you've done for macOS. We could say it's always kernel 3.14, for example.

colelao replied:
You're right. I looked into the spec, and it says if platform is Linux then version should be an empty string. I'll fix this now

mcatanzaro commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -40,31 +56,116 @@ Ref<NavigatorUAData> NavigatorUAData::create()
+    return true;
+#else
     return false;
+#endif
```
We can do better here:

```
#if USE(GLIB)
#include <wtf/glib/ChassisType.h>
#endif
```

```suggestion
#if PLATFORM(IOS_FAMILY)
    return true;
#elif USE(GLIB)
    return chassisType() == WTF::ChassisType::Mobile;
#else
    return false;
#endif
```

colelao replied:
This is great. I had no idea that existed

mcatanzaro commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -40,31 +56,116 @@ Ref<NavigatorUAData> NavigatorUAData::create()
+            values->architecture = "x86"_s;
+#else
+            values->architecture = "arm"_s;
+#endif
```
Are you intentionally lying about the architecture? That seems fine to me, but we should add a comment to explain why we're doing this. ("Websites might discriminate against the user if we provide a truthful answer.") Otherwise, somebody will likely come along in the future and change this to respond truthfully using `utsname.machine`.

(Also, "x86" and "arm" are unusual ways to write the architecture names. I see these are not part of the client hints spec. Are you trying to match what Chrome does?)

colelao replied:
I was following this part of the spec:
>[User agents](https://infra.spec.whatwg.org/#user-agent) MUST map higher-entropy [platform architecture](https://wicg.github.io/ua-client-hints/#user-agent-platform-architecture) values to the following buckets:
>
>x86 CPU architectures => "x86"
>ARM CPU architectures => "arm"

I'll add a comment.

karlcow replied:
If it's not sent by the current UA string, it would be probably better to not send them at all.

mcatanzaro replied:
`platformVersionForUAString` from UserAgentGLib.cpp effectively exposes `uname -m`, so this is actually a little less information.

mcatanzaro reviewed (commented):
Can you look into https://github.com/WebKit/standards-positions/issues/70 please? Our standards position is marked as "blocked" which isn't true if we are implementing it.

brentfulgham commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -40,31 +60,108 @@ Ref<NavigatorUAData> NavigatorUAData::create()
 bool NavigatorUAData::mobile() const
 {
+#if PLATFORM(IOS_FAMILY)
+    return true;
```
Now that Brandon has landed his change, I think we should use something like:

```
#if PLATFORM(IOS_FAMILY)
    return !(PAL::currentUserInterfaceIdiomIsDesktop() || PAL::currentUserInterfaceIdiomIsVision());
#else ...
```

brentfulgham replied:
Note: We also need to take into account cases where the user has selected the "Request Desktop Site" option for the page. We should do something here so that we match that user choice.

brentfulgham commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -40,31 +60,108 @@ Ref<NavigatorUAData> NavigatorUAData::create()
+    });
+    return platformName->isolatedCopy();
+#elif PLATFORM(IOS_FAMILY)
+    return "iOS"_s;
```
This is wrong for iPad. We report ourselves as macOS on that platform (unless the user requests the mobile site).

So you should probably do something like:
```
#elif PLATFORM(IOS_FAMILY)
    return (PAL::currentUserInterfaceIdiomIsDesktop() || PAL::currentUserInterfaceIdiomIsVision()) ? "macOS"_s : "iOS"_s;
#else
   ...    
```

brentfulgham reviewed (changes requested):
I think this is looking good. Please apply the changes Michael, Karl, and I suggested and I think we can land this. We should file a follow-up Bugzilla to make sure that this API respects the user's ability to change the User Agent (e.g., in the Safari Developer menu). If someone changes the UA for loading a site, this API should report the same things.

brentfulgham commented on Source/WebCore/page/UADataValues.h:
```diff
@@ -42,5 +43,14 @@ struct UADataValues {
     String uaFullVersion;
     bool wow64;
+
+    static Ref<UADataValues> create(Vector<NavigatorUABrandVersion> brands, bool mobile, String platform)
```
`brands` should be passed as a `const Vector<NavigatorUABrandVersion>&`, and `platform` as a `String&&`.

brentfulgham commented on Source/WebCore/page/UADataValues.h:
```diff
@@ -42,5 +43,14 @@ struct UADataValues {
+    }
+
+private:
+    UADataValues(Vector<NavigatorUABrandVersion> brands, bool mobile, String platform)
```
Ditto the data types for `brands` and `platform` from the `create` function.

brentfulgham commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -40,31 +61,108 @@ Ref<NavigatorUAData> NavigatorUAData::create()
+
+        auto rng = std::default_random_engine { };
+        std::ranges::shuffle(temp, rng);
+        brandVersion.construct(temp);
```
What is the purpose of this shuffle operation?

brentfulgham replied:
Is the spec such that we return a list of brands, and we want to randomize the set returned from one WCP launch to another?

colelao replied:
I was following this part of the spec. I wasn't sure if it meant every launch or not, so I went with every time to be safe. 

>The [user agent](https://infra.spec.whatwg.org/#user-agent) SHOULD execute the following steps:
>1. [Append](https://infra.spec.whatwg.org/#list-append) one additional [item](https://infra.spec.whatwg.org/#list-item) to list containing a [NavigatorUABrandVersion](https://wicg.github.io/ua-client-hints/#dictdef-navigatoruabrandversion) dictionary, initialized with [brand](https://wicg.github.io/ua-client-hints/#dom-navigatoruabrandversion-brand) set to [arbitrary brand](https://wicg.github.io/ua-client-hints/#create-an-arbitrary-brand) and set [version](https://wicg.github.io/ua-client-hints/#dom-navigatoruabrandversion-version) to the result of [create an arbitrary version](https://wicg.github.io/ua-client-hints/#create-an-arbitrary-version) with version type.
>2. Randomize the order of the [items](https://infra.spec.whatwg.org/#list-item) in list.

brentfulgham replied:
Ah! Then that's perfect.

brentfulgham commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -40,31 +61,108 @@ Ref<NavigatorUAData> NavigatorUAData::create()
+
+String NavigatorUAData::createArbitraryVersion()
+{
+    return makeString( weakRandomNumber<unsigned>() % 10000, '.', weakRandomNumber<unsigned>() % 10000, '.', weakRandomNumber<unsigned>() % 10000);
```
Nit: No whitespace between `makeString(` and `weakRandomNumber`, please.

brentfulgham reviewed (approved):
Thanks for cleaning up those last few items. r=me.

colelao committed:
Implement User-Agent Client Hints - navigator.userAgentData
https://bugs.webkit.org/show_bug.cgi?id=241749
rdar://95454627

Reviewed by Brent Fulgham.

This patch populates the userAgentData data structure to be used for quirking. It follows this spec:
https://wicg.github.io/ua-client-hints/#getHighEntropyValues

Sensitive information is denied such as model and formFactors.

* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::userAgentData const):
(WebCore::Navigator::initializeNavigatorUAData const): Deleted.
* Source/WebCore/page/Navigator.h:
* Source/WebCore/page/NavigatorUAData.cpp:
(WebCore::NavigatorUAData::brands const):
(WebCore::NavigatorUAData::mobile const):
(WebCore::NavigatorUAData::platform const):
(WebCore::NavigatorUAData::toJSON const):
(WebCore::NavigatorUAData::getHighEntropyValues const):
(WebCore::NavigatorUAData::createArbitraryVersion):
(WebCore::NavigatorUAData::createArbitraryBrand):
* Source/WebCore/page/NavigatorUAData.h:
* Source/WebCore/page/UADataValues.h:
(WebCore::UADataValues::create):
(WebCore::UADataValues::UADataValues):
* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::userAgentData const):
(WebCore::WorkerNavigator::initializeNavigatorUAData const): Deleted.
* Source/WebCore/page/WorkerNavigator.h:

Canonical link: https://commits.webkit.org/298242@main
